### PR TITLE
Update react-query type definition for v1

### PR DESCRIPTION
**Pull request by lukaszfiszer**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/tannerlinsley/react-query/blob/master/CHANGELOG.md#100
- [X] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by lukaszfiszer** [comment]
Lot's of code style errors - will fix them soon

File: types/react-query/index.d.ts
Comment (purge) [inline types/react-query/index.d.ts]: this should just be TResult I think

File: types/react-query/index.d.ts
Comment (purge) [inline types/react-query/index.d.ts]: this should be `TResult` I think

File: types/react-query/index.d.ts
Comment (purge) [inline types/react-query/index.d.ts]: why not `| void`

File: types/react-query/index.d.ts
Comment (purge) [inline types/react-query/index.d.ts]: most of these functions have optional params, i.e `exact?:` and `throwOnError?:`

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: https://github.com/Microsoft/dtslint/blob/master/docs/void-return.md

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Yes, good catch

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Indeed, will fix

**Comment by purge** [comment]
@lukaszfiszer what do you think about a discriminated union on `status` thus allowing `data` to not be undefined in that state?

**Comment by lukaszfiszer** [comment]
@purge you mean discriminated union of QueryResult interface on status and date fields?

File: types/react-query/index.d.ts
Comment (yss14) [inline types/react-query/index.d.ts]: Shouldn't this be `refetchAllOnWindowFocus?: boolean;`?
[Docs](https://github.com/tannerlinsley/react-query#window-focus-refetching) says this property can be set to `false`.

**Review by Fiszcz — APPROVED** [review]

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Fixed

**Comment by lukaszfiszer** [comment]
@purge - discriminated union implemented

**Comment by purge** [comment]
@lukaszfiszer nice! i checked the source earlier and it does appear that data can still be set for states other than 'success' so you might wanna make that an `|`

**Comment by lukaszfiszer** [comment]
@purge you mean that data should be `data: TResult | undefined` for error and loading status, but `data: TResult` for success status? Does it apply to paginated and infinite query?

**Comment by purge** [comment]
@lukaszfiszer i believe that to be the case, yes (you may want to continue showing data while loading)

File: types/react-query/index.d.ts
Comment (Fiszcz) [inline types/react-query/index.d.ts]: I would add ``undefined`` to this union.

What do you think ?

Comment (Fiszcz) [inline types/react-query/index.d.ts]: I would add ``undefined`` to this union.

What do you think ?

**Review by Fiszcz — APPROVED** [review]

File: types/react-query/index.d.ts
Comment (purge) [inline types/react-query/index.d.ts]: these should be | void

**Comment by yss14** [comment]
Will [optional variables](ICreateRankingPayload) support for queries be added in a subsequent PR?

File: types/react-query/index.d.ts
Comment (JaceHensley) [reply types/react-query/index.d.ts]: Should it be `Promise<void> | void`?

File: types/react-query/index.d.ts
Comment (rbuckton) [inline types/react-query/index.d.ts]: This type of generic definition is unsafe as there is nothing that guards the return value to ensure it is a `TResult` (which is why there is a lint error on this line). It might be better to return `unknown` here and have the caller of the function perform an appropriate cast or type guard.

File: types/react-query/index.d.ts
Comment (rbuckton) [inline types/react-query/index.d.ts]:
```suggestion
    function getQuery(queryKey: QueryKey): unknown;
```

File: types/react-query/index.d.ts
Comment (rbuckton) [inline types/react-query/index.d.ts]:
```suggestion
    function subscribe(callback: (queryCache: unknown) => void): (unsubscribe: (...args: any[]) => void) => void;
```

**Review by rbuckton — CHANGES_REQUESTED** [review]
@lukaszfiszer can you respond to the other reviewers comments as well?

File: types/react-query/index.d.ts
Comment (Oscarteg) [inline types/react-query/index.d.ts]:
```suggestion
    onError?: (error: Error) => void;
```
This should have the error as parameter and not return an object but void I think

File: types/react-query/index.d.ts
Comment (Oscarteg) [inline types/react-query/index.d.ts]:
```suggestion
    onSuccess?: (data: object) => data;
```
This should have data as the parameter and return the same data. I'm not sure how we can type the data.

File: types/react-query/index.d.ts
Comment (Oscarteg) [inline types/react-query/index.d.ts]:
```suggestion
    onSettled?: (data: object, error: Error) => data;
```

**Review by oscarteg — COMMENTED** [review]
Changes to the ReactQueryProviderConfig interface to reflect the correct `onError`, `onSuccess` and `onSettled` API.

File: types/react-query/index.d.ts
Comment (Dema) [reply types/react-query/index.d.ts]: I think, onSuccess/onSettled `data` param should have type of `TResult`.

Comment (Oscarteg) [reply types/react-query/index.d.ts]: > I think, onSuccess/onSettled `data` param should have type of `TResult`.

@dema I don't know how that is possible because you would have to set the generic in advance of using it. 

I'm not an expert on Typescript though.

Comment (Dema) [reply types/react-query/index.d.ts]: Yeah, my bad, I didn't notice that it's about ProviderConfig.

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: ok

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: > `void` should be used as a return type, but not as a parameter type.

https://github.com/Microsoft/dtslint/blob/master/docs/void-return.md

Comment (Oscarteg) [reply types/react-query/index.d.ts]:
```suggestion
    onSuccess?: (data: TResult) => Promise<any> | undefined;
```

Comment (JaceHensley) [reply types/react-query/index.d.ts]: I think that would count as a return type. It looks like that means func param not generic param

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Isn't letting the caller perform the cast even more error-prone in this case? Even when they cast correctly to type associated with this query key, they may still forget to check for undefined value. Compare:

```
type IData = { foo: boolean};
const data = getQueryData<IData>('key'); // data is IData or undefined
data.foo // TS2532: Object is possibly 'undefined'.
```

vs
 
```
const data = getQueryData('key') as IData; 
data.foo // no error
```

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: @JaceHensley you're right. I've mistook that with other line of code. will fix

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: `onSuccess` callback must not return the same data - even the default implementation is a `noop` fn (https://github.com/tannerlinsley/react-query/blob/master/src/config.js#L19).  

```
onSuccess?: (data: unknown) => Promise<any> | void
```
should be the correct signature

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Will change, but making both arguments optional

Comment (Oscarteg) [reply types/react-query/index.d.ts]: > Will change, but making both arguments optional

Good call. They should be optional.

**Comment by lukaszfiszer** [comment]
Everyone thanks for you feedback!  @Oscarteg @Fiszcz @rbuckton  i've updated PR with your suggestions

@purge it is stated explicitely in docs that loading state is for initial fetch without any cached data. (https://github.com/tannerlinsley/react-query#returns) Do you have an example/a code pen with data when state is loading?

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Changed the code followign your suggestion, but still explanation is much appreciated :-)

**Comment by lukaszfiszer** [comment]
@rbuckton can you mark the changes you'd requested as resolved? Otherwise it won't be merged I guess

File: types/react-query/index.d.ts
Comment (yss14) [inline types/react-query/index.d.ts]: Why `TVariables extends object`? Can't `variables` also be a primitive type like `string` or `number`?

Comment (purge) [reply types/react-query/index.d.ts]: it should be a generic with a reasonable default

Comment (Oscarteg) [reply types/react-query/index.d.ts]: And also the variables are always passed as object to the mutation function so it has to extend object.

Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Definition of `useMutation` was mostly copy-pasted from v.0, but I guess it need some attention too!

File: types/react-query/index.d.ts
Comment (burtyish) [inline types/react-query/index.d.ts]: missing `reset` function
https://github.com/tannerlinsley/react-query#resetting-mutation-state
```suggestion
    data: TResults | undefined;
    reset: () = void;
```

Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Thanks. Its also worth raising an issue in react-query repo to fix the docs, which misses this `reset` function

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Fixed. However, I had trouble correctly typing first argument of mutate function - see TODOs in test file. Any help would be appreciated

File: types/react-query/index.d.ts
Comment (simlrh) [inline types/react-query/index.d.ts]: Do these hooks need overloads? According to the docs the second argument can be an array of additional parameters to `queryFn` https://github.com/tannerlinsley/react-query#optional-variables

Comment (simlrh) [reply types/react-query/index.d.ts]: Also can be a single object argument https://github.com/tannerlinsley/react-query#using-a-query-object-instead-of-parameters

**Comment by smashercosmo** [comment]
New version of react-query has just been released 
```
1. Added the ability to use staleTime: Infinity for queries that should never go stale
2. Added the queryCache.getQueries function
3. useMutation can now allow multiple mutation requests at the same
```

**Comment by lukaszfiszer** [comment]
@smashercosmo 
1. `Infinity` is of type `number` - no changes needed in types
2. type definition `queryCache.getQueries` added
3. no change in types requied

**Comment by lukaszfiszer** [comment]
@rbuckton - can you review and approve?

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: They definitely do need overloads, however as they might quite difficult to type correctly I planned to add them later on after this initial PR is merged

**Comment by nucleartux** [comment]
I have two problems with latest typings with `useInfiniteQuery`: 
1. 
image
`getFetchMore` result expects to be equal `QueryKey` but it should not, `getFetchMore` just adds third optional parameter to `queryFn` (https://github.com/tannerlinsley/react-query#load-more--infinite-scroll-with-useinfinitequery)
2. 
image
`data` still has type `[] | TResult[]` even if I check `status === 'success``. So I can't really iterate over `data` without manually casting types.

**Comment by lukaszfiszer** [comment]
@nucleartux you're right. The docs were not very clear on that. I took some time trying to fix the types, but found it quite difficult. It would probably required a rewrite a good part of `useInfiniteQuery` types, introducing a new generic parameter `TQueryFunctionArgs` in place of `TQueryKey`. The latter does not have any information about the type that needs to be returned from `getFetchMore`

File: types/react-query/index.d.ts
Comment (smashercosmo) [inline types/react-query/index.d.ts]: initialData can be a function

**Review by smashercosmo — CHANGES_REQUESTED** [review]

File: types/react-query/index.d.ts
Comment (smashercosmo) [inline types/react-query/index.d.ts]: useQuery can accept object as argument (same for usePaginatedQuery and useInfiniteQuery)
```js
useQuery({
  queryKey,
  queryFn,
  variables,
  config
})
```

**Review by smashercosmo — CHANGES_REQUESTED** [review]

File: types/react-query/index.d.ts
Comment (clementgarbay) [inline types/react-query/index.d.ts]: Shouldn't this be?
```suggestion
    getFetchMore?: (lastGroup: TResult, allGroups: TResult[]) => FetchMoreVariable | boolean;
```

From what I understand in the [doc](https://github.com/tannerlinsley/react-query#options-2) and the code it's the result of the function that is either a `FetchMoreVariable` or a `boolean`.

Maybe this could be the right typing:
```ts
    getFetchMore?: (lastGroup: TResult, allGroups: TResult[]) => FetchMoreVariable | false;
```

**Comment by bjarkehs** [comment]
At this point I think it would be better to get something merged, and then tweak it later.

**Comment by simlrh** [comment]
IMHO an incomplete initial implementation is fine, but it shouldn't flag errors on correct code, which these types currently do.

**Comment by barrymay** [comment]
Just curious, I know the rbuckton approval appears to be blocking this - but he hasn’t been around to confirm the changes. What about creating a new PR and getting that re-approved if that’s the blocker?

 (I’m not debating anyone else’s more active concerns, but in these times you never know why someone’s away)

File: types/react-query/index.d.ts
Comment (Fiszcz) [inline types/react-query/index.d.ts]: Yo Łukasz,

What do you think about this: 
```typescript
export function useQuery<TResult, TQueryKey extends QueryKey>(
    queryKey: TQueryKey | (() => TQueryKey),
    queryFn: QueryFunction<TResult, TVariables>,	    queryFn: (key: string, ...args: QueryKeyVariablesArgs<TQueryKey>) => Promise<TResult>,
    options?: QueryOptions<TResult>	    options?: QueryOptions<TResult>,
): QueryResult<TResult, TVariables>;	): QueryState<TResult>;

export function useQuery<TResult, TQueryKey extends QueryKey, TQueryVariables extends any[]>(
    queryKey: TQueryKey | (() => TQueryKey),
    queryVariables: TQueryVariables,
    queryFn: QueryFunction<TResult, TVariables>,	    queryFn: (key: string, ...args: Concat<QueryKeyVariablesArgs<TQueryKey>, TQueryVariables>) => Promise<TResult>,
    options?: QueryOptions<TResult>	    options?: QueryOptions<TResult>,
): QueryResult<TResult, TVariables>;	): QueryState<TResult>;
```
?

This is simple solution for optional queryVariables, it's not perfect, but for first version of types for *React-Query* should be fine.

``Concat<T, R>`` is implementation from **typescript-tuple** package.

**Review by Fiszcz — APPROVED** [review]

**Comment by nucleartux** [comment]
Also currently I can't pass falsy value to queryKey because my fetchFn still expects arguments from from non falsy situation.  Of course I can do something about this myself but anyway it don't work as expected.

**Review by Igorbek — CHANGES_REQUESTED** [review]
I mistakenly have been working on the same problem wich resulted in my PR #43438

I think I made type inference for functions, keys, and variables to work quite well. So I checked the tests I used with this PR and found a few issues that you might want to add to your cases as well.

```ts
const queryFn = (name: string, params: { bar: string }) => Promise.resolve(10);

declare const condition: boolean;

// Query with falsey query key
useQuery(condition && ['foo', { bar: 'baz' }], queryFn);
//                                             ~~~~~~~
// Argument of type '(name: string, params: { bar: string; }) => Promise<number>' is not assignable to parameter of type '(key: string, ...args: [] | [{ bar: string; }]) => Promise<number>'.
//   Types of parameters 'params' and 'args' are incompatible.
//     Type '[] | [{ bar: string; }]' is not assignable to type '[{ bar: string; }]'.
//       Property '0' is missing in type '[]' but required in type '[{ bar: string; }]'.

// This is just missing object API (when all the parameters passed as an object)
useQuery({
    queryKey: condition && ['foo', { bar: 'baz' }],
    queryFn,
});
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Expected 2-3 arguments, but got 1.
// An argument for 'queryFn' was not provided.

const query2 = useQuery(
    ['key', { a: 1 }],
    [{ b: true }, { c: 'c' }],
    //~~~~~~~~~~~~~~~~~~~~~~~
    // Argument of type '({ b: boolean; } | { c: string; })[]' is not assignable to parameter of type '(key: string, args_1: { a: number; }) => Promise<unknown>'.
    //   Type '({ b: boolean; } | { c: string; })[]' provides no match for the signature '(key: string, args_1: { a: number; }): Promise<unknown>'.
    //
    (
        key1,
        //~~ Parameter 'key1' implicitly has an 'any' type.
        key2,
        //~~ Parameter 'key2' implicitly has an 'any' type.
        var1,
        //~~ Parameter 'var1' implicitly has an 'any' type.
        var2,
        //~~ Parameter 'var2' implicitly has an 'any' type.
    ) => Promise.resolve(key1 === 'key' && key2.a === 1 && var1.b === true && var2.c === 'c'),
);
query2.data; // $ExpectType undefined | boolean
//     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Expected: undefined | boolean; got: unknown
```

**Comment by lukaszfiszer** [comment]
Since @Igorbek provided a more complete implementation in #43438, I will close this PR. Please take your time and review there.

**Pull request closed without merging**
